lib/src/highlighter.dart

```diff
@@ -44,28 +47,496 @@ class Highlighter {
-    newSpan = _normalizeTrailingNewline(newSpan);
-    newSpan = _normalizeEndOfLine(newSpan);
-
-    return Highlighter._(newSpan, color as String);
+  /// be used to highlight [span]'s text (for example, `"\u001b[31m"` will color
+  /// red). If it's `true`, it indicates that the text should be highlighted
+  /// using the default color. If it's `false` or `null`, it indicates that no
+  /// color should be used.
+  Highlighter(SourceSpan span, {color})
+      : this._(_collateLines([_Highlight(span, primary: true)]), () {
+          if (color == true) return colors.red;
+          if (color is String) return color;
+          return null;
+        }(), null);
+
+  /// Creates a [Highlighter] that will return a string highlighting
+  /// [primarySpan] as well as all the spans in [secondarySpans] within the text
+  /// of their file when [highlight] is called.
+  ///
+  /// All spans are assumed to be part of the same single source file. Each one
+  /// has an associated label that will be written alongside it. For
+  /// [primarySpan] this message is [primaryLabel], and for [secondarySpans] the
+  /// labels are the map values.
+  ///
+  /// If [color] is `true`, this will use [ANSI terminal color
+  /// escapes](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors) to
+  /// highlight the text. The [primarySpan] will be highlighted with
+  /// [primaryColor] (which defaults to red), and the [secondarySpans] will be
+  /// highlighted with [secondaryColor] (which defaults to blue). These
+  /// arguments are ignored if [color] is `false`.
+  Highlighter.multiple(SourceSpan primarySpan, String primaryLabel,
+      Map<SourceSpan, String> secondarySpans,
+      {bool color = false, String primaryColor, String secondaryColor})
+      : this._(
+            _collateLines([
+              _Highlight(primarySpan, label: primaryLabel, primary: true),
+              for (var entry in secondarySpans.entries)
+                _Highlight(entry.key, label: entry.value)
+            ]),
+            color ? primaryColor ?? colors.red : null,
+            color ? secondaryColor ?? colors.blue : null);
+
+  Highlighter._(this._lines, this._primaryColor, this._secondaryColor)
+      : _paddingBeforeSidebar = 1 +
+            math.max<int>(
+                // In a purely mathematical world, floor(log10(n)) would give the
+                // number of digits in n, but floating point errors render that
+                // unreliable in practice.
+                (_lines.last.number + 1).toString().length,
+                // If [_lines] aren't contiguous, we'll write "..." in place of a
+                // line number.
+                _contiguous(_lines) ? 0 : 3),
+        _maxMultilineSpans = _lines
+            .map((line) => line.highlights
+                .where((highlight) => isMultiline(highlight.span))
+                .length)
+            .fold(0, math.max),
+        _multipleFiles = !isAllTheSame(_lines.map((line) => line.url));
```

Comment:
The `isAllTheSame` method might be simpler using `every`, or negated `any`.


How about

```dart
_multipleFiles = _lines.isNotEmpty && _lines.any((l) => l.url != _lines.first.url);
```